### PR TITLE
pay gas with address balances - TransactionExpiration modeling and validation

### PR DESCRIPTION
**Pull request by alex-mysten**
## Description 

Add new TransactionExpiration definition and associated validation. Users specify an empty payment vec on TransactionData; address balance will be used.

## Test plan 

UTs and e2e tests added

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] gRPC:
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+        _ => panic!("Expected MissingGasPayment error when feature flag is disabled"),
+    }
+}
+
+#[test]
+fn test_address_balance_payment_with_accumulators_enabled() {
+    let _guard = ProtocolConfig::apply_overrides_for_testing(|_, mut cfg| {
+        cfg.enable_address_balance_gas_payments_for_testing();
+        cfg
+    });
```
Comment (mystenmark) [inline crates/sui-types/src/unit_tests/address_balance_gas_tests.rs]: Note - since you are supplying the protocol config directly, you can just create the config and call the setters on it without using the override hook. That's only needed for cases where you have no way to inject a config (like when using TestClusterBuilder).

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+            min_epoch: Some(1),
+            max_epoch: Some(1),
+            min_timestamp_seconds: None,
+            max_timestamp_seconds: None,
+            chain: ChainIdentifier::from(CheckpointDigest::default()),
+            nonce: 123,
+        },
+    );
+
+    let result = tx_data.validity_check(&config);
```
Comment (mystenmark) [inline crates/sui-types/src/unit_tests/address_balance_gas_tests.rs:107]: how is it getting the current epoch?

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+        UserInputError::MissingTransactionExpiration => {}
+        _ => panic!("Expected MissingTransactionExpiration error"),
+    }
+
+    let tx_data = create_test_transaction_data(vec![], TransactionExpiration::Epoch(1));
+
+    let result = tx_data.validity_check(&config);
+    assert!(result.is_err());
+    match result.unwrap_err() {
+        UserInputError::MissingTransactionExpiration => {}
```
Comment (mystenmark) [inline crates/sui-types/src/unit_tests/address_balance_gas_tests.rs]: i wonder if we need a better error that communicates that `TransactionExpiration::Epoch` is not sufficient here. It does appear to be an expiration, so I could see someone being very confused here.

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+            assert!(msg.contains("Both min_epoch and max_epoch must be specified and equal"));
+        }
+        _ => panic!("Expected Unsupported error for missing epochs"),
+    }
+
+    let tx_data = create_test_transaction_data(
+        vec![],
+        TransactionExpiration::ValidDuring {
+            min_epoch: Some(1),
+            max_epoch: None,
```
Comment (mystenmark) [inline crates/sui-types/src/unit_tests/address_balance_gas_tests.rs:241]: should we also have a test for `min_epoch: None, max_epoch: Some(...)`

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+            chain: ChainIdentifier::from(CheckpointDigest::default()),
+            nonce: 123,
+        },
+    );
+
+    let result = tx_data.validity_check(&config);
+    assert!(result.is_err());
+    match result.unwrap_err() {
+        UserInputError::Unsupported(msg) => {
+            assert!(msg.contains("Multi-epoch transaction expiration is not yet supported"));
```
Comment (mystenmark) [inline crates/sui-types/src/unit_tests/address_balance_gas_tests.rs:346]: we could actually support this easily for any transaction with an owned input.

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+            chain: ChainIdentifier::from(CheckpointDigest::default()),
+            nonce: 123,
+        },
+    );
+
+    let result = tx_data.validity_check(&config);
+    assert!(result.is_err());
+    match result.unwrap_err() {
+        UserInputError::Unsupported(msg) => {
+            assert!(msg.contains("Both min_epoch and max_epoch must be specified and equal"));
```
Comment (mystenmark) [inline crates/sui-types/src/unit_tests/address_balance_gas_tests.rs:372]: similar comment here

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+
+#[test]
+fn test_address_balance_payment_requires_valid_during_only() {
+    let _guard = ProtocolConfig::apply_overrides_for_testing(|_, mut cfg| {
+        cfg.enable_address_balance_gas_payments_for_testing();
+        cfg
+    });
+    let config = ProtocolConfig::get_for_max_version_UNSAFE();
+
+    let tx_data = create_test_transaction_data(vec![], TransactionExpiration::None);
```
Comment (mystenmark) [inline crates/sui-types/src/unit_tests/address_balance_gas_tests.rs]: i think you already test this case on line 131?

File: crates/sui-types/src/transaction.rs [hunk 8]
```diff
@@ -1723,13 +1727,28 @@ pub struct GasData {
...
+}
+
 #[derive(Debug, PartialEq, Eq, Hash, Clone, Copy, Serialize, Deserialize)]
 pub enum TransactionExpiration {
     /// The transaction has no expiration
     None,
     /// Validators wont sign a transaction unless the expiration Epoch
     /// is greater than or equal to the current epoch
     Epoch(EpochId),
+    /// Valid during the times specified
```
Comment (mystenmark) [inline crates/sui-types/src/transaction.rs]: Let's have a more extensive comment explaining the motivation / design of this variant

File: crates/sui-types/src/error.rs [hunk 9]
```diff
@@ -285,6 +285,17 @@ pub enum UserInputError {
...
+    #[error("Invalid wait for effects request: {error}")]
+    InvalidWaitForEffectsRequest { error: String },
+
+    #[error("Transaction with empty gas payment must have expiration")]
+    MissingTransactionExpiration,
+
+    #[error(
+        "Transaction expiration chain ID {provided} does not match network chain ID {expected}"
+    )]
+    InvalidExpirationChainId { provided: String, expected: String },
```
Comment (mystenmark) [inline crates/sui-types/src/error.rs]: Maybe just call this `InvalidChainId`?

File: crates/sui-types/src/transaction.rs [hunk 10]
```diff
@@ -2740,10 +2803,40 @@ impl SenderSignedData {
...
+                        context.epoch > *max
+                    } else {
+                        false
+                    }
+                } else if let Some(max) = max_epoch {
+                    context.epoch > *max
+                } else {
+                    false
+                }
+            }
```
Comment (mystenmark) [inline crates/sui-types/src/transaction.rs]: i find the `if match { /* stuff that yields true if there's an error */ } { error }` to be a bit confusing. Can we just have a match that returns errors for the bad cases?

We can also handle both the expiration and chain id checks in the same place then

File: crates/sui-types/src/transaction.rs [hunk 11]
```diff
@@ -1723,13 +1727,28 @@ pub struct GasData {
...
     None,
     /// Validators wont sign a transaction unless the expiration Epoch
     /// is greater than or equal to the current epoch
     Epoch(EpochId),
+    /// Valid during the times specified
+    ValidDuring {
+        min_epoch: Option<EpochId>,
+        max_epoch: Option<EpochId>, // inclusive
+        min_timestamp_seconds: Option<u64>,
+        max_timestamp_seconds: Option<u64>, // inclusive
```
Comment (lxfind) [inline crates/sui-types/src/transaction.rs]: These two options will be tricky to support universally. For one it won't work with fast-path transactions.
So we are introducing a dependency from expiration format to requiring consensus for the transaction, which is a bit unintuitive.

File: crates/sui-types/src/transaction.rs [hunk 12]
```diff
@@ -1723,13 +1727,28 @@ pub struct GasData {
...
+        chain: ChainIdentifier,
+        nonce: u32,
```
Comment (lxfind) [inline crates/sui-types/src/transaction.rs:1775]: Could you doc comment why we need chain and nonce here?

File: crates/sui-types/src/transaction.rs (hunk 11 above)
Comment (mystenmark) [reply crates/sui-types/src/transaction.rs]: It would be easy to support, in that it is yet another "flag" in the transaction (like the presence of a withdraw) that would force the transaction to wait for consensus. If we enable this in the future, we'll have to communicate that clearly.

There are no immediate plans to enable this, but its an obvious enough feature that I think its good to have this here so we don't have to evolve the format in the future

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs (hunk 2 above)
Comment (alex-mysten) [reply crates/sui-types/src/unit_tests/address_balance_gas_tests.rs:107]: There are two `validity_check`s, one is a static validity check and one depends on validator state. These UTs only exercise static validations; we rely on e2e tests to exercise stateful validations

[crates/sui-types/src/transaction.rs#L2725](https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/transaction.rs#L2725)

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs (hunk 5 above)
Comment (alex-mysten) [reply crates/sui-types/src/unit_tests/address_balance_gas_tests.rs:346]: It seems to me that would make the API difficult to reason about. I'm happy to revisit this if you think it's important

File: crates/sui-indexer-alt-graphql/src/api/types/transaction/mod.rs [hunk 13]
```diff
@@ -114,6 +114,13 @@ impl TransactionContents {
             TransactionExpiration::Epoch(epoch_id) => {
                 Ok(Some(Epoch::with_id(self.scope.clone(), *epoch_id)))
             }
+            TransactionExpiration::ValidDuring { max_epoch, .. } => {
+                if let Some(epoch_id) = max_epoch {
+                    Ok(Some(Epoch::with_id(self.scope.clone(), *epoch_id)))
```
Comment (mystenmark) [inline crates/sui-indexer-alt-graphql/src/api/types/transaction/mod.rs:116]: @amnn can you take a look at this? we probably need a better API here because the existing one cannot really capture the new information

Comment (amnn) [reply crates/sui-indexer-alt-graphql/src/api/types/transaction/mod.rs:116]: Sorry for missing this -- is the issue that you `max_epoch` is potentially in the future? If so, it's not technically a problem -- it's fine to return a reference to an `Epoch` in the future -- it just means that you can only fetch the `epochId` field on that (until that epoch is no longer in the future, at which point you can fetch other details about it).

Comment (alex-mysten) [reply crates/sui-indexer-alt-graphql/src/api/types/transaction/mod.rs:116]: I think the concern is that the API might behave surprisingly if expiration could be determined on additional parameters besides epoch; timestamp for example. Correct me if I'm wrong @mystenmark

Comment (amnn) [reply crates/sui-indexer-alt-graphql/src/api/types/transaction/mod.rs:116]: Ah I see, no problem, if you can give me a sketch of the various ways that transaction validity can be described, then I can show how to express that in GraphQL. We can mark this field as deprecated and create a new field and type that has a more expressive type.

Comment (alex-mysten) [reply crates/sui-indexer-alt-graphql/src/api/types/transaction/mod.rs:116]: Seems reasonable. A 1:1 type mapping with the TransactionExpiration variants seems more appropriate to me and ensures correctness and future-proofing.

I'm not sure we should block landing this on creating a more expressive API to return expiration. The current implementation is the most sensible backwards-compatible behavior. max_epoch is required even if there are additional constraints, so the API would return correct but partial data.

What do you think?

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs (hunk 1 above)
Comment (alex-mysten) [reply crates/sui-types/src/unit_tests/address_balance_gas_tests.rs]: Addressed

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs (hunk 3 above)
Comment (alex-mysten) [reply crates/sui-types/src/unit_tests/address_balance_gas_tests.rs]: Addressed

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs (hunk 4 above)
Comment (alex-mysten) [reply crates/sui-types/src/unit_tests/address_balance_gas_tests.rs:241]: Addressed

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs (hunk 6 above)
Comment (alex-mysten) [reply crates/sui-types/src/unit_tests/address_balance_gas_tests.rs:372]: Addressed

File: crates/sui-types/src/unit_tests/address_balance_gas_tests.rs (hunk 7 above)
Comment (alex-mysten) [reply crates/sui-types/src/unit_tests/address_balance_gas_tests.rs]: Addressed

File: crates/sui-types/src/transaction.rs (hunk 8 above)
Comment (alex-mysten) [reply crates/sui-types/src/transaction.rs]: Addressed

File: crates/sui-types/src/error.rs (hunk 9 above)
Comment (alex-mysten) [reply crates/sui-types/src/error.rs]: Addressed

File: crates/sui-types/src/transaction.rs (hunk 10 above)
Comment (alex-mysten) [reply crates/sui-types/src/transaction.rs]: Addressed

File: crates/sui-types/src/transaction.rs (hunk 12 above)
Comment (alex-mysten) [reply crates/sui-types/src/transaction.rs:1775]: Addressed

File: crates/sui-e2e-tests/tests/address_balance_tests.rs [hunk 14]
```diff
@@ -361,3 +401,386 @@ fn make_send_to_account_tx(
...
+                error_message.contains("GasBalanceTooLow"),
+                "Expected GasBalanceTooLow because gas payments with address balance not implemented, but got: {}",
+                error_message
+            );
+        }
+        Ok((effects, _)) => {
+            assert!(
+                effects.status().is_err(),
+                "Expected transaction to fail due to unimplemented charge mechanics, but got success"
+            );
```
Comment (mystenmark) [inline crates/sui-e2e-tests/tests/address_balance_tests.rs]: i'm confused why there are two error cases here - shouldn't we deterministically hit only one of them?

File: crates/sui-e2e-tests/tests/address_balance_tests.rs [hunk 15]
```diff
@@ -361,3 +401,386 @@ fn make_send_to_account_tx(
...
+
+    let signed_tx = test_cluster.sign_transaction(&tx).await;
+    let result = test_cluster
+        .execute_transaction_return_raw_effects(signed_tx)
+        .await;
+
+    let err = result.expect_err("Transaction should be rejected when only max_epoch is specified");
+    let err_str = format!("{:?}", err);
+    assert!(
+        err_str.contains("Both min_epoch and max_epoch must be specified and equal"),
```
Comment (mystenmark) [inline crates/sui-e2e-tests/tests/address_balance_tests.rs:690]: If i'm not mistaken, this transaction uses a gas coin, so it should be able to specify any epoch range it wants to.

We could be stricter than necessary and leave this as is, since I don't think anybody is asking for that functionality, but at least leave a TODO comment

Comment (alex-mysten) [reply crates/sui-e2e-tests/tests/address_balance_tests.rs:690]: Yep, I know it's overly strict and not against relaxing the rules down the road. Added a todo

File: crates/sui-e2e-tests/tests/address_balance_tests.rs (hunk 14 above)
Comment (alex-mysten) [reply crates/sui-e2e-tests/tests/address_balance_tests.rs]: Fixed
